fix: refactor integration test infra, re-enable `testproject-tools-integration`, fix flaky tests on desktop & console platforms

NoelStephensUnity opened the pull request:
This PR migrates the runtime integration test helper classes into their own assembly in order to improve upon collaboration and standardization of writing Netcode for GameObject integration tests.  This also includes additional modifications for some tests (to leverage from new features) in order to get them passing on all non-Desktop platforms.

This also includes an overhaul on the NetcodeIntegrationTest (formerly known as BaseMultiInstanceTest) that provides more functionality/flexibility depending upon your preferred style of writing integration tests.

[MTT-2435](https://jira.unity3d.com/browse/MTT-2435)

## Testing and Documentation
* Includes updates to unit tests.
* [Integration Test Updates Document](https://docs.google.com/document/d/1jwVKqTDSVdokKtno6T6UwZBEmLWznbrNxM7xkBUrhQw/edit#heading=h.avo638yzegpz)

## Console Build:
This PR also has been consistently passing console builds.

NoelStephensUnity committed:
chore

Create the new com.unity.netcode.testhelpers assembly.
Migrate SIPTransport into the new assembly.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
chore

Migrate MultiInstanceHelpers and BaseMultiInstanceTest into com.unity.netcode.testhelpers

NoelStephensUnity committed:
fix

Updating the NGO assembly.

NoelStephensUnity committed:
fix

excluding unet if it is not present.

NoelStephensUnity committed:
update

Migrated NetworkManagerHelper and NetworkVariableHelper into Unity.Netcode.TestHelpers.

Updated tests that used these classes to use the Unity.Netcode.TestHelpers namespace.

NoelStephensUnity committed:
chore

Renaming MultiInstance to NetcodeIntegration.
So:
BaseMultiinstanceTest -> NetcodeIntegrationTest
MultiInstanceHelpers -> NetcodeIntegrationTestHelpers

This will probably change as I migrate some of NetcodeIntegrationTestHelpers into NetcodeIntegrationTest.

NoelStephensUnity committed:
update

Fixes for UNetTransport exclusion and added additional support for UTP in command line handler.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Removed all references to NetcodeIntegrationTestHelpers.Run within tests.
Preparing NetcodeIntegrationTest.WaitForConditionOrTimeOut to be used in placed of the NetcodeIntegrationTestHelpers.WaitForCondition

NoelStephensUnity committed:
fix

Missed a few references to NetcodeIntegrationTestHelpers.Run

NoelStephensUnity committed:
update

Testing only runtime references to this assembly and adding UnityEngine.TestRunner and NUnit library.

NoelStephensUnity committed:
chore

renaming the assembly to better clarify that these test helpers are only for runtime integration tests.

NoelStephensUnity committed:
update

This includes the core changes to NetworkSceneManager, integration test helper classes, and the scene handler integration test.

NoelStephensUnity committed:
test

Updated all of the rest of the tests that were modified in PR-1405

NoelStephensUnity committed:
update

Missed an update in NetworkScenemanagerTests

NoelStephensUnity committed:
Merge branch 'fix/networkscenemanager-interface-override-backport-develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Merging #1691 into this branch ahead of it being merged to continue working on mtt-2435 while it waits to get approved.

NoelStephensUnity committed:
chore

Migrate multiplayer tools helper classes out of Unity.Netcode.Runtime assembly and into the Unity.Netcode.TestHelpers.Runtime assembly.

NoelStephensUnity committed:
style

Fixed whitespace issue.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update and fix

Fixed the issue with two versions of IntegrationTestSceneHandler.
Updated the NetcodeIntegrationTest to be able to run all tests contained within a derived class using the same NetworkManager instances (i.e. doesn't spin up a NetworkManager per UnityTest)

NoelStephensUnity committed:
Update

Added additional functionality to the setup and teardown process.
Minor addition to creating a player prefab that doesn't require passing an action to the StartSomeClientsAndServerWithPlayers method (which will be broken up into phases).

NoelStephensUnity committed:
test and update

Migrating most of the editor serialization test methods into runtime tests.

NoelStephensUnity committed:
Update

Added pre and post teardown virtual methods.
Fixed issues with the tools project SceneEventTests (still need to look through this more to make sure each scenario is being tested properly).

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
fix

Fixed issue with merge.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Extracting the TimeOutHelper and ConditionalPredicate stuff into their own files.
Added better functionality to NetworkObject destruction that allows the developer to select what gets destroyed and what is kept around between integration tests.

NoelStephensUnity committed:
style

fixing white space

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
fix

Wrong place to clean up NetworkObjects.

NoelStephensUnity committed:
update

This breaks apart the server and client creation from the server and client starting as well as provides a way to know the server and client(s) are created (not started but created).

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
fix

Fixing new test that was brought into the develop branch.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Migrated wait for clients into NetcodeIntegrationTest (still kept the NetCodeIntegrationTestHelpers version).
Removed the couroutine references in NetCodeIntegrationTestHelpers and renamed the CoroutineResultsWrapper to just ResultsWrapper.
Updated some comments in  NetcodeIntegrationTest.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Final modifications to the IntegrationTestHelper class.  This should provide many more options and flexibility when writing an integration test.
Updated a few classes with the modifications made.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
fix

Updating NetworkObjectSpawnManyObjectsTests to new integration test class names.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Finalized the updates and made some adjustments to classes that would demonstrate additional features of the new NetcodeIntegrationTest class.
Update comments and cleaned up a bit.

Other than keeping up with new tests that are merged, this should be the last of the changes.

NoelStephensUnity committed:
Fix

Removing Debug.Break

NoelStephensUnity committed:
style

Just cleaning up the comments a bit.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Migrating a merged experimental-develop over to use NetcodeIntegrationTest.
While doing so, I noticed some redundancy that was not needed in the setup and teardown virtual methods.  Revised this area of NetcodeIntegrationTest.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Removed duplicate test entry in NetworkVariableTests (it was already backported in the first round of console test fixes).

Updated NetworkVisibility test to handle running the server as a host and as just a server.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
style

Updating comments and renamed some methods in NetcodeIntegrationTest for clarity.

NoelStephensUnity committed:
update

The following update were changes needed to get the tools tests working when running in both the editor and as a stand alone test runner build.
This also includes minor updates to comments, some tests that needed minor tweaking, and re-enables the testing of the tools integration tests.

NoelStephensUnity committed:
update and style

Renamed some NetcodeIntegrationTest methods for clarity
Trying to figure out the issue with the tools tests not compiling.

NoelStephensUnity committed:
revert

Reverting some of the project.metafile changes from last update.

NoelStephensUnity committed:
revert

Re-applying the exclusion of the testproject-tools-integration for Yamato builds until we can figure out the best plan of action to handling the compilation issues only specific to this project.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Updated the player naming.
Updated the newly added NetworkObjectNetworkClientOwnedObjectsTests to derive from NetcodeIntegrationTest

NoelStephensUnity committed:
style

white space fix

NoelStephensUnity committed:
update

Created a more global way to identify spawned NetworkObjects within integration tests.
During this time, I discovered some minor issues with tracking the player objects and resolved that.
Also discovered that if you add a NetworkBehaviour to the player object and try to run NetworkObjectDestroyTests.TestNetworkObjectClientDestroy it will throw an exception because the NetworkBehaviourUpdater continues to try to run the NetworkBehaviour even though it no longer exists (i.e. the NetworkObject.ChildNetworkBehaviours will still contain NetworkBehaviours that no longer exist).
While the issue still exists, I made adjustments to the ObjectNameIdentifier that takes this into consideration until it is determined to be a bug or expected.

NoelStephensUnity committed:
update

Putting back the NetworkObjectSpawnManyObjectsTests client and spawn count that I was using for other testing.

NoelStephensUnity committed:
update

Added some additional helper methods to NetcodeIntegrationTest

becksebenius-unity commented on .yamato/project.metafile:
```diff
@@ -50,9 +50,9 @@ projects:
-#      - trunk
+  #- name: testproject-tools-integration
+  #  path: testproject-tools-integration
+  #  validate: true
```
Not sure I understand this change, if the job is still commented out, was the change from false to true necessary?

NoelStephensUnity replied:
No that will get reverted out... good catch.

becksebenius-unity commented on com.unity.netcode.gameobjects/Tests/Editor/com.unity.netcode.editortests.asmdef:
```diff
@@ -28,12 +27,5 @@
         "UNITY_INCLUDE_TESTS"
     ],
     "versionDefines": [],
-    "noEngineReferences": false,
-    "versionDefines": [
-        {
-            "name": "com.unity.multiplayer.tools",
-            "expression": "",
-            "define": "MULTIPLAYER_TOOLS"
-        }
-    ]
```
Why did you remove this? Did you test NGO tests without tools package installed? The purpose of this is to make sure that NGO tests still compile and can be run by users who don't have tools package installed.

NoelStephensUnity replied:
Let me look into this...

NoelStephensUnity replied:
Yeah, that was definitely a good catch.  I think I was comparing it to the testproject editor tests assembly definition.
Either case it is fixed.

becksebenius-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/HiddenVariableTests.cs:
```diff
@@ -41,30 +42,29 @@ public void ListChanged(NetworkListEvent<int> listEvent)
-                    var networkTransform = playerPrefab.AddComponent<HiddenVariableTest>();
-                    m_TestNetworkPrefab = PreparePrefab();
-                });
+            var networkTransform = m_PlayerPrefab.AddComponent<HiddenVariableTest>();
```
A little strange, why save to a local variable if it's never used?

NoelStephensUnity replied:
Ahh...that is actually an artifact from the previous code... there is no point to that and I will go ahead and remove that local var.

NoelStephensUnity committed:
update

Adding the integration test examples used in the documentation.

NoelStephensUnity committed:
fix

Updating the NetworkVisbilityTest

NoelStephensUnity committed:
update

reverting the validate for testproject-tools-integration back to false.

NoelStephensUnity committed:
fix and style

Fixing issue with spawning with ownership and removing local variable from HiddenVariableTests

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Updating the NetworkAnimatorTests to use the NetcodeIntegrationTest

NoelStephensUnity committed:
revert and fix

Reverting accidental removal of MULTIPLAYER_TOOLS define.
Fixing issue with NetworkAnimatorTests referencing the editor version of AnimatorController.

NoelStephensUnity committed:
style

white space fix

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Reduce the player NetworkObjects down to one dictionary as the server-side was redundant.
Updated tests to reflect the change.

NoelStephensUnity committed:
update manifest

updating to the 1.0.0-pre.5 multiplayer tools package.

NoelStephensUnity committed:
update

Updated some of the metrics tests to changes in NetcodeIntegrationTest.

NoelStephensUnity committed:
update

uncommenting the tools testproject

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Merge updates
Converted RpcTests

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

0xFA11 committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update and style

This is just migrating the message hook conditional and RpcMessageHooks classes into their own files.

NoelStephensUnity committed:
style and test

Updated comments in NetcodeIntegrationTest.
Added the last document example updated to IntegrationTestExamples.

0xFA11 commented on .yamato/project.metafile:
```diff
@@ -45,9 +45,9 @@ projects:
-#    test_editors:
-#      - 2021.2
-#      - trunk
+  - name: testproject-tools-integration
+    path: testproject-tools-integration
+    validate: false
+    test_editors:
+      - 2021.2
+      - trunk
```
this looks out of scope in this PR

becksebenius-unity replied:
I asked for this - this was turned off recently because tests were unstable due to missing fixes from experimental-develop, and this refactor/backport in theory was supposed to fix it

0xFA11 replied:
nope, those changes got reverted entirely by #1743

becksebenius-unity replied:
can you elaborate? I don't see the relationship with that PR

I'm not opposed to re-enabling the tools integration tests into a separate PR though

0xFA11 replied:
nope I'm stupid, I misread these as multiprocess for some reason.
but still looks out of scope, there's too much going on under this PR.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Editor/com.unity.netcode.editortests.asmdef:
```diff
@@ -9,10 +9,9 @@
-        "TestAssemblies"
+        "Unity.Multiplayer.Tools.NetStats",
+        "UnityEngine.TestRunner",
+        "UnityEditor.TestRunner"
```
doesn't `TestAssemlies` cover `UnityEngine.TestRunner` & `UnityEditor.TestRunner` already? (in a safer way)

NoelStephensUnity replied:
Yeah, I was trying to keep it consistent with the rest of the asmdef files that do it this way.
Reverted that out and keeping the optionalUnityReferences

NoelStephensUnity replied:
Actually, if you use the previous way it causes issues in newer versions of unity.
The above is what is applied by default if you use the optionalUnityReferences... by including that it creates duplicate constraint entries for UNITY_INCLUDE_TESTS and it creates duplicate nunit.framework.dll assembly references.  When you open the asmdef in the editor, remove the duplicate entries, and then hit apply the above is what you get.

0xFA11 commented on com.unity.netcode.gameobjects/Utilities/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -0,0 +1,681 @@
+    /// <summary>
+    /// The default Netcode For GameObjects integration test helper class
+    /// </summary>
+    public abstract class NetcodeIntegrationTest
```
one problem I have with this PR is that it does multiple things which is too much.
for instance, I'm looking at MTT-2435's title & description: "Can not test individual test assemblies without RuntimeTests assembly"
this PR is doing much more than that, it does rename & refactor too, even some CI changes and so on.
I think we need to split this PR into more focused PR chunks and get them in separately rather than in a one single big lump sum.

NoelStephensUnity replied:
We definitely should discuss this as splitting this up will cause issues.
I was actually going to rename MTT-2435 to be better suited to the changes.
This PR also ended up fixing the issues with [PR-1665 ](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/pull/1665) as well as making sure that all tests back ported from experimental-develop to develop passed on consoles.

We can split it up, but it could become complicated at this point.

0xFA11 replied:
can't we start with new asmdef without any fixes, renames or any other changes first?
later, we could work on renames then fixes specifically, no?

NoelStephensUnity replied:
We could... but the longer we wait the more things will need to be updated and the more integration tests will be written the legacy way (which means more things to update).  
But we can definitely break this all out over a span of time for sure.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/IntegrationTestExamples.cs:
```diff
@@ -0,0 +1,192 @@
+    public class IntegrationTestUpdated : NetcodeIntegrationTest
+    {
+        private GameObject m_MyNetworkPrefab;
+        protected override int NbClients => 1;
```
can we just rename `NbClients` to `NumberOfClients` please?

NoelStephensUnity replied:
Yeah, we can do that.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/IntegrationTestExamples.cs:
```diff
@@ -0,0 +1,192 @@
+        {
+            // Check the condition for this test and automatically handle varying processing
+            // environments and conditions
+            yield return WaitForConditionOrTimeOut(() =>
+            Object.FindObjectsOfType<NetworkVisibilityComponent>().Where(
+                (c) => c.IsSpawned).Count() == 2);
+
+            Assert.False(s_GloabalTimeOutHelper.TimedOut, "Timed out waiting for instances " +
+                "to be detected!");
```
is this file a new addition or a refactor/rename?

```suggestion
            yield return WaitForConditionOrTimeOut(() =>
            Object.FindObjectsOfType<NetworkVisibilityComponent>().Where(c => c.IsSpawned).Count() == 2);

            Assert.False(s_GloabalTimeOutHelper.TimedOut, "Timed out waiting for instances to be detected!");
```

NoelStephensUnity replied:
This file contains the examples within the document for reference purposes.  I can exclude it but figured it could be useful to have as reference and to see that the examples actually work.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/IntegrationTestExamples.cs:
```diff
@@ -0,0 +1,192 @@
+            // Check the condition for this test and automatically handle varying processing
+            // environments and conditions
+            yield return WaitForConditionOrTimeOut(() =>
+            Object.FindObjectsOfType<NetworkVisibilityComponent>().Where(
+                (c) => c.IsSpawned).Count() == 2);
+
+            Assert.False(s_GloabalTimeOutHelper.TimedOut, "Timed out waiting for instances " +
+                "to be detected!");
```
nit/minor style:
```suggestion
            Object.FindObjectsOfType<NetworkVisibilityComponent>().Where(c => c.IsSpawned).Count() == 2);

            Assert.False(s_GloabalTimeOutHelper.TimedOut, "Timed out waiting for instances to be detected!");
```

NoelStephensUnity replied:
Yeah, I had formatted these examples to better fit in the companion document.

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/IntegrationTestExamples.cs:
```diff
@@ -0,0 +1,192 @@
+            var expected = m_NumberToSpawn * TotalClients;
+            // Wait for all clients to have spawned all instances
+            yield return WaitForConditionOrTimeOut(() => SpawnTest.TotalSpawned == expected);
+            Assert.False(s_GloabalTimeOutHelper.TimedOut, $"Timed out waiting for all to " +
+                $"spawn! Total Spawned: {SpawnTest.TotalSpawned}");
+
+            var client1Relative = s_GlobalNetworkObjects[1].Values.Where((c) =>
+            c.gameObject.GetComponent<SpawnTest>() != null);
+            foreach (var networkObject in client1Relative)
+            {
+                var testComp = networkObject.GetComponent<SpawnTest>();
+                // Confirm each one is owned by the server
+                Assert.IsTrue(testComp.IsOwnedByServer, $"{testComp.name} is not owned" +
+                    $" by the server!");
+            }
+        }
+
+        [UnityTest]
+        [Order(2)]
+        public IEnumerator TestDespawnNetworkObjects()
+        {
+            var serverRelative = s_GlobalNetworkObjects[0].Values.Where((c) =>
+            c.gameObject.GetComponent<SpawnTest>() != null).ToList();
+            foreach (var networkObject in serverRelative)
+            {
+                networkObject.Despawn();
+            }
+            // Wait for all clients to have spawned all instances
+            yield return WaitForConditionOrTimeOut(() => SpawnTest.TotalSpawned == 0);
+            Assert.False(s_GloabalTimeOutHelper.TimedOut, $"Timed out waiting for all to " +
+                $"despawn! Total Spawned: {SpawnTest.TotalSpawned}");
```
nit/minor style:
```suggestion
            Assert.False(s_GloabalTimeOutHelper.TimedOut, $"Timed out waiting for all to spawn! Total Spawned: {SpawnTest.TotalSpawned}");

            var client1Relative = s_GlobalNetworkObjects[1].Values.Where((c) =>
            c.gameObject.GetComponent<SpawnTest>() != null);
            foreach (var networkObject in client1Relative)
            {
                var testComp = networkObject.GetComponent<SpawnTest>();
                // Confirm each one is owned by the server
                Assert.IsTrue(testComp.IsOwnedByServer, $"{testComp.name} is not owned by the server!");
            }
        }

        [UnityTest]
        [Order(2)]
        public IEnumerator TestDespawnNetworkObjects()
        {
            var serverRelative = s_GlobalNetworkObjects[0].Values.Where((c) =>
            c.gameObject.GetComponent<SpawnTest>() != null).ToList();
            foreach (var networkObject in serverRelative)
            {
                networkObject.Despawn();
            }
            // Wait for all clients to have spawned all instances
            yield return WaitForConditionOrTimeOut(() => SpawnTest.TotalSpawned == 0);
            Assert.False(s_GloabalTimeOutHelper.TimedOut, $"Timed out waiting for all to despawn! Total Spawned: {SpawnTest.TotalSpawned}");
```

NoelStephensUnity replied:
Same thing regarding fitting within a document for readability.
I can remove this, but then the examples won't match the document examples (will be the same code...just without the restriction on line columns consumed).

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs:
```diff
@@ -57,7 +46,7 @@ public IEnumerator TrackNetworkObjectSpawnSentMetric()
             var objectSpawned = objectSpawnedSentMetricValues.Last();
             Assert.AreEqual(Client.LocalClientId, objectSpawned.Connection.Id);
-            Assert.AreEqual($"{k_NewNetworkObjectName}(Clone)", objectSpawned.NetworkId.Name);
+            Assert.AreEqual($"{spawnedObject.name}", objectSpawned.NetworkId.Name);
```
```suggestion
            Assert.AreEqual(spawnedObject.name, objectSpawned.NetworkId.Name);
```

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs:
```diff
@@ -67,16 +56,17 @@ public IEnumerator TrackNetworkObjectSpawnReceivedMetric()
             Assert.AreEqual(Server.LocalClientId, objectSpawned.Connection.Id);
             Assert.AreEqual(networkObject.NetworkObjectId, objectSpawned.NetworkId.NetworkId);
-            Assert.AreEqual($"{k_NewNetworkObjectName}(Clone)", objectSpawned.NetworkId.Name);
+            Assert.AreEqual($"{clientSideObject.name}", objectSpawned.NetworkId.Name);
```
```suggestion
            Assert.AreEqual(clientSideObject.name, objectSpawned.NetworkId.Name);
```

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs:
```diff
@@ -85,20 +75,22 @@ public IEnumerator TrackNetworkObjectDestroySentMetric()
             var objectDestroyed = objectDestroyedSentMetricValues.Last();
             Assert.AreEqual(Client.LocalClientId, objectDestroyed.Connection.Id);
-            Assert.AreEqual($"{k_NewNetworkObjectName}(Clone)", objectDestroyed.NetworkId.Name);
+            Assert.AreEqual($"{objectName}", objectDestroyed.NetworkId.Name);
```
```suggestion
            Assert.AreEqual(objectName, objectDestroyed.NetworkId.Name);
```

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs:
```diff
@@ -120,8 +115,8 @@ public IEnumerator TrackNetworkObjectDestroyReceivedMetric()
-            Assert.AreEqual(networkObject.NetworkObjectId, objectDestroyed.NetworkId.NetworkId);
-            Assert.AreEqual($"{k_NewNetworkObjectName}(Clone)", objectDestroyed.NetworkId.Name);
+            Assert.AreEqual(objectId, objectDestroyed.NetworkId.NetworkId);
+            Assert.AreEqual($"{objectName}", objectDestroyed.NetworkId.Name);
```
```suggestion
            Assert.AreEqual(objectName, objectDestroyed.NetworkId.Name);
```

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/com.unity.netcode.runtimetests.asmdef:
```diff
@@ -11,11 +11,12 @@
     ],
     "includePlatforms": [],
     "excludePlatforms": [],
-    "allowUnsafeCode": false,
+    "allowUnsafeCode": true,
```
what's the unsafe code getting into this asmdef now?

NoelStephensUnity replied:
Because of the serialization and fastbuffer related tests.

0xFA11 commented on com.unity.netcode.gameobjects/Utilities/TestHelpers/Runtime/NetworkVariableHelper.cs:
```diff
@@ -1,7 +1,7 @@
 using System.Collections.Generic;
 
-namespace Unity.Netcode.RuntimeTests
+namespace Unity.Netcode.TestHelpers.Runtime
```
I feel like we should just use `Unity.Netcode.TestHelpers` namespace for all and also drop `.Runtime`

NoelStephensUnity replied:
This was planning for any potential editor specific helpers and trying to keep it clear that these are runtime test only helper methods.
I can flatten it down, but it opens the possibility of someone adding editor specific helpers in this assembly (which would/could cause issues).

0xFA11 commented on testproject-tools-integration/Packages/manifest.json:
```diff
@@ -1,10 +1,11 @@
-    "com.unity.ide.rider": "3.0.7",
-    "com.unity.netcode.gameobjects": "file:../../com.unity.netcode.gameobjects",
-    "com.unity.multiplayer.tools": "1.0.0-pre.4",
+    "com.unity.ide.rider": "3.0.9",
+    "com.unity.ide.visualstudio": "2.0.14",
+    "com.unity.multiplayer.tools": "1.0.0-pre.5",
     "com.unity.netcode.adapter.utp": "file:../../com.unity.netcode.adapter.utp",
+    "com.unity.netcode.gameobjects": "file:../../com.unity.netcode.gameobjects",
```
we probably don't need changes in `manifest.json`

NoelStephensUnity replied:
Yes we do...
Take note of this:  "com.unity.multiplayer.tools": "1.0.0-pre.5"

0xFA11 replied:
so, we only need `"com.unity.multiplayer.tools": "1.0.0-pre.4"` to `"com.unity.multiplayer.tools": "1.0.0-pre.5"` change then? :)

NoelStephensUnity replied:
The only other changes were:
- keeping the "com.unity.ide.visualstudio": "2.0.14" in place so anyone using visual studio wouldn't need to have to keep adding the package locally.
- keeping the updated rider package version for those that use rider.
I made sure to get approval from Beck before these two additions, but if you think they should be reverted then I am happy to do that.

0xFA11 reviewed (changes requested):
I think this PR needs to be broken down into multiple separate PRs, each focusing on one single goal.

NoelStephensUnity committed:
style

renamed NBClients to NumberOfClients

NoelStephensUnity committed:
update

Fixing compile error due to name change of NBClients to NumberOfClients

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
last update

Had merged this before realizing this PR might need to be broken up.  This preserves the original integration test as well as the changes to get the valid portion of this integration test working.

NoelStephensUnity committed:
style

removing using System

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/Metrics/NetworkObjectMetricsTests.cs

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
update

using optionalUnityReferences.

NoelStephensUnity committed:
fix

Removed using system from the wrong file.

NoelStephensUnity committed:
fix-revert

Reverting the suggested change.

NoelStephensUnity committed:
fix

disabling the Transform interpolation test as it is not stable and needs to be re-thought through.

0xFA11 commented on com.unity.netcode.gameobjects/Utilities/TestHelpers/Runtime/Components/ObjectNameIdentifier.cs:
```diff
@@ -0,0 +1,91 @@
+        private bool m_IsRegistered;
+
+        // Keep a reference to the assigned NetworkObject
+        // <see
```
👀

0xFA11 commented on com.unity.netcode.gameobjects/Utilities/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -0,0 +1,681 @@
+    /// </summary>
+    public abstract class NetcodeIntegrationTest
+    {
+        static protected TimeOutHelper s_GloabalTimeOutHelper = new TimeOutHelper(4.0f);
+        static protected WaitForSeconds s_DefaultWaitForTick = new WaitForSeconds(1.0f / k_DefaultTickRate);
```
```suggestion
        protected static TimeoutHelper s_GloabalTimeoutHelper = new TimeoutHelper(4.0f);
        protected static WaitForSeconds s_DefaultWaitForTick = new WaitForSeconds(1.0f / k_DefaultTickRate);
```

0xFA11 replied:
some nits :P

NoelStephensUnity committed:
update

minimal conversion of TransformInterpolationTests to new NetcodeIntegrationTest.

NoelStephensUnity committed:
update

Adding host or server selection capabilities to reduce code replication for this common task.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
update

Merge updates.
Refactored the RpcMessageHooks to be more generic and renamed it to MessageHooksConditional.  It allows for (n) number of message types to be assigned to (x) number of clients. Updated the tests that utilize the message hooks.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-2435-and-integration-test-cleanup

NoelStephensUnity committed:
style

renaming TimeOutHelper to TimeoutHelper.

NoelStephensUnity committed:
style

finishing the thought on the m_NetworkObject comment

NoelStephensUnity committed:
style

white space fix

NoelStephensUnity committed:
style

removed system namespace.

0xFA11 committed:
minor update

0xFA11 committed:
move serialization unit tests back into editortests

0xFA11 committed:
update `testproject-tools-integration`'s `manifest.json`

0xFA11 committed:
cleanup asmdefs

0xFA11 committed:
update projects & packages

0xFA11 committed:
update `manifest.json`
